File: backend/app/schemas/insights.py
"""Pydantic schemas for climate, agriculture, aviation, and city insights."""

from datetime import datetime
from typing import Optional

from pydantic import BaseModel, Field

from app.schemas.weather import GeoLocation


# --------------------------------------------------------------------- #
# Historical / climate trends
# --------------------------------------------------------------------- #


class HistoricalDailyPoint(BaseModel):
    """One day of historical (archive) weather observations."""

    date: str
    temperature_max: Optional[float] = None
    temperature_min: Optional[float] = None
    temperature_mean: Optional[float] = None
    precipitation_sum: Optional[float] = None
    wind_speed_max: Optional[float] = None


class HistoricalPeriodStats(BaseModel):
    """Aggregate statistics for the requested historical window."""

    period_days: int
    temp_mean: Optional[float] = None
    temp_max: Optional[float] = None
    temp_min: Optional[float] = None
    total_precipitation: Optional[float] = None
    wet_days: int = 0  # days with >= 1 mm precipitation
    hottest_day: Optional[HistoricalDailyPoint] = None
    wettest_day: Optional[HistoricalDailyPoint] = None


class HistoricalWeatherResponse(BaseModel):
    """GET /weather/historical response."""

    location: GeoLocation
    start_date: str
    end_date: str
    daily: list[HistoricalDailyPoint] = []
    stats: HistoricalPeriodStats


class ClimateTrendPoint(BaseModel):
    """One month of aggregated climate data."""

    month: str  # e.g. "2025-01"
    avg_temp_max: Optional[float] = None
    avg_temp_min: Optional[float] = None
    avg_temp_mean: Optional[float] = None
    total_precipitation: Optional[float] = None
    wet_days: int = 0


class ClimateTrendResponse(BaseModel):
    """GET /weather/climate response — monthly aggregates over N years."""

    location: GeoLocation
    years: int
    monthly: list[ClimateTrendPoint] = []
    # Simple linear warming trend in °C per decade (None if insufficient data).
    warming_trend_c_per_decade: Optional[float] = None
    # Long-term annual precipitation average (mm).
    annual_precipitation_mm: Optional[float] = None
    # Same calendar month from previous years, for "is this month unusual?".
    current_month_anomaly_c: Optional[float] = None


# --------------------------------------------------------------------- #
# Crop advisories (agriculture decision support)
# --------------------------------------------------------------------- #


class CropAdvisoryRequest(BaseModel):
    """POST /weather/crop-advisory request body."""

    latitude: float = Field(..., ge=-90, le=90)
    longitude: float = Field(..., ge=-180, le=180)
    location_name: Optional[str] = Field(None, max_length=120)
    crop: Optional[str] = Field(
        None, max_length=40, description="Optional crop key (rice, wheat, cotton, …)"
    )


class CropAdvisory(BaseModel):
    """One actionable advisory bullet."""

    category: str  # irrigation | pest_disease | field_work | sowing | general
    severity: str  # info | caution | warning
    message: str


class CropAdvisoryResponse(BaseModel):
    """POST /weather/crop-advisory response."""

    location: str
    latitude: float
    longitude: float
    crop: Optional[str] = None
    advisories: list[CropAdvisory] = []
    # 0-100 suitability for field operations (spraying, harvesting, ploughing)
    field_work_score: Optional[int] = None
    # Growing-degree-day style heat accumulation over the forecast window
    irrigation_needed: Optional[bool] = None
    generated_at: datetime = Field(default_factory=datetime.now)


# --------------------------------------------------------------------- #
# Aviation briefing
# --------------------------------------------------------------------- #


class AviationCondition(BaseModel):
    """A single aviation-relevant observation/rule result."""

    parameter: str  # visibility | wind | ceiling | precipitation | storm | temp
    status: str  # ok | caution | hazard
    value: Optional[str] = None
    note: Optional[str] = None


class AviationBriefingRequest(BaseModel):
    """POST /weather/aviation request body."""

    latitude: float = Field(..., ge=-90, le=90)
    longitude: float = Field(..., ge=-180, le=180)
    location_name: Optional[str] = Field(None, max_length=120)
    hours_ahead: int = Field(12, ge=1, le=24)


class AviationBriefingResponse(BaseModel):
    """POST /weather/aviation response."""

    location: str
    # ok | caution | hazard — worst of the individual conditions
    flight_category: str
    summary: str
    conditions: list[AviationCondition] = []
    # VFR-style window: hour ranges where conditions are acceptable
    best_windows: list[str] = []
    generated_at: datetime = Field(default_factory=datetime.now)


# --------------------------------------------------------------------- #
# Smart-city multi-location overview
# --------------------------------------------------------------------- #


class CityWeatherSnapshot(BaseModel):
    """Current weather snapshot for one city."""

    name: str
    latitude: float
    longitude: float
    temperature: Optional[float] = None
    feels_like: Optional[float] = None
    humidity: Optional[float] = None
    wind_speed: Optional[float] = None
    precipitation: Optional[float] = None
    condition: Optional[str] = None
    weather_code: Optional[int] = None
    aqi: Optional[int] = None


class CityOverviewResponse(BaseModel):
    """GET /weather/city-overview response — monitor many cities at once."""

    cities: list[CityWeatherSnapshot] = []
    generated_at: datetime = Field(default_factory=datetime.now)

File: mobile/flutter_app/lib/models/insights.dart
// Models for insights features: climate history, crop advisories,
// aviation briefings, and city overview.

// ---------------- Climate / historical ----------------

class HistoricalStats {
  final int periodDays;
  final double? tempMean;
  final double? tempMax;
  final double? tempMin;
  final double? totalPrecipitation;
  final int wetDays;
  final String? hottestDayDate;
  final double? hottestDayTemp;
  final String? wettestDayDate;
  final double? wettestDayPrecip;

  const HistoricalStats({
    required this.periodDays,
    this.tempMean,
    this.tempMax,
    this.tempMin,
    this.totalPrecipitation,
    this.wetDays = 0,
    this.hottestDayDate,
    this.hottestDayTemp,
    this.wettestDayDate,
    this.wettestDayPrecip,
  });

  factory HistoricalStats.fromJson(Map<String, dynamic> json) => HistoricalStats(
        periodDays: (json['period_days'] as num?)?.toInt() ?? 0,
        tempMean: (json['temp_mean'] as num?)?.toDouble(),
        tempMax: (json['temp_max'] as num?)?.toDouble(),
        tempMin: (json['temp_min'] as num?)?.toDouble(),
        totalPrecipitation: (json['total_precipitation'] as num?)?.toDouble(),
        wetDays: (json['wet_days'] as num?)?.toInt() ?? 0,
        hottestDayDate:
            (json['hottest_day'] as Map<String, dynamic>?)?['date'] as String?,
        hottestDayTemp: (json['hottest_day'] as Map<String, dynamic>?) == null
            ? null
            : ((json['hottest_day'] as Map<String, dynamic>)['temperature_max']
                    as num?)
                ?.toDouble(),
        wettestDayDate:
            (json['wettest_day'] as Map<String, dynamic>?)?['date'] as String?,
        wettestDayPrecip:
            (json['wettest_day'] as Map<String, dynamic>?) == null
                ? null
                : ((json['wettest_day'] as Map<String, dynamic>)['precipitation_sum']
                        as num?)
                    ?.toDouble(),
      );
}

class HistoricalDaily {
  final String date;
  final double? temperatureMax;
  final double? temperatureMin;
  final double? temperatureMean;
  final double? precipitationSum;

  const HistoricalDaily({
    required this.date,
    this.temperatureMax,
    this.temperatureMin,
    this.temperatureMean,
    this.precipitationSum,
  });

  factory HistoricalDaily.fromJson(Map<String, dynamic> json) => HistoricalDaily(
        date: json['date'] as String? ?? '',
        temperatureMax: (json['temperature_max'] as num?)?.toDouble(),
        temperatureMin: (json['temperature_min'] as num?)?.toDouble(),
        temperatureMean: (json['temperature_mean'] as num?)?.toDouble(),
        precipitationSum: (json['precipitation_sum'] as num?)?.toDouble(),
      );
}

class HistoricalWeather {
  final String startDate;
  final String endDate;
  final List<HistoricalDaily> daily;
  final HistoricalStats stats;

  const HistoricalWeather({
    required this.startDate,
    required this.endDate,
    required this.daily,
    required this.stats,
  });

  factory HistoricalWeather.fromJson(Map<String, dynamic> json) =>
      HistoricalWeather(
        startDate: json['start_date'] as String? ?? '',
        endDate: json['end_date'] as String? ?? '',
        daily: (json['daily'] as List<dynamic>?)
                ?.map((e) => HistoricalDaily.fromJson(e as Map<String, dynamic>))
                .toList() ??
            [],
        stats: HistoricalStats.fromJson(
            json['stats'] as Map<String, dynamic>? ?? {}),
      );
}

class ClimateTrendMonth {
  final String month;
  final double? avgTempMax;
  final double? avgTempMin;
  final double? avgTempMean;
  final double? totalPrecipitation;
  final int wetDays;

  const ClimateTrendMonth({
    required this.month,
    this.avgTempMax,
    this.avgTempMin,
    this.avgTempMean,
    this.totalPrecipitation,
    this.wetDays = 0,
  });

  factory ClimateTrendMonth.fromJson(Map<String, dynamic> json) =>
      ClimateTrendMonth(
        month: json['month'] as String? ?? '',
        avgTempMax: (json['avg_temp_max'] as num?)?.toDouble(),
        avgTempMin: (json['avg_temp_min'] as num?)?.toDouble(),
        avgTempMean: (json['avg_temp_mean'] as num?)?.toDouble(),
        totalPrecipitation: (json['total_precipitation'] as num?)?.toDouble(),
        wetDays: (json['wet_days'] as num?)?.toInt() ?? 0,
      );
}

class ClimateTrend {
  final int years;
  final List<ClimateTrendMonth> monthly;
  final double? warmingTrendCPerDecade;
  final double? annualPrecipitationMm;
  final double? currentMonthAnomalyC;

  const ClimateTrend({
    required this.years,
    required this.monthly,
    this.warmingTrendCPerDecade,
    this.annualPrecipitationMm,
    this.currentMonthAnomalyC,
  });

  factory ClimateTrend.fromJson(Map<String, dynamic> json) => ClimateTrend(
        years: (json['years'] as num?)?.toInt() ?? 0,
        monthly: (json['monthly'] as List<dynamic>?)
                ?.map((e) =>
                    ClimateTrendMonth.fromJson(e as Map<String, dynamic>))
                .toList() ??
            [],
        warmingTrendCPerDecade:
            (json['warming_trend_c_per_decade'] as num?)?.toDouble(),
        annualPrecipitationMm:
            (json['annual_precipitation_mm'] as num?)?.toDouble(),
        currentMonthAnomalyC:
            (json['current_month_anomaly_c'] as num?)?.toDouble(),
      );
}

// ---------------- Crop advisories ----------------

class CropAdvisoryItem {
  final String category;
  final String severity;
  final String message;

  const CropAdvisoryItem({
    required this.category,
    required this.severity,
    required this.message,
  });

  factory CropAdvisoryItem.fromJson(Map<String, dynamic> json) =>
      CropAdvisoryItem(
        category: json['category'] as String? ?? 'general',
        severity: json['severity'] as String? ?? 'info',
        message: json['message'] as String? ?? '',
      );
}

class CropAdvisory {
  final String location;
  final String? crop;
  final List<CropAdvisoryItem> advisories;
  final int? fieldWorkScore;
  final bool? irrigationNeeded;

  const CropAdvisory({
    required this.location,
    this.crop,
    required this.advisories,
    this.fieldWorkScore,
    this.irrigationNeeded,
  });

  factory CropAdvisory.fromJson(Map<String, dynamic> json) => CropAdvisory(
        location: json['location'] as String? ?? 'Unknown',
        crop: json['crop'] as String?,
        advisories: (json['advisories'] as List<dynamic>?)
                ?.map((e) =>
                    CropAdvisoryItem.fromJson(e as Map<String, dynamic>))
                .toList() ??
            [],
        fieldWorkScore: (json['field_work_score'] as num?)?.toInt(),
        irrigationNeeded: json['irrigation_needed'] as bool?,
      );
}

// ---------------- Aviation briefing ----------------

class AviationConditionItem {
  final String parameter;
  final String status;
  final String? value;
  final String? note;

  const AviationConditionItem({
    required this.parameter,
    required this.status,
    this.value,
    this.note,
  });

  factory AviationConditionItem.fromJson(Map<String, dynamic> json) =>
      AviationConditionItem(
        parameter: json['parameter'] as String? ?? '',
        status: json['status'] as String? ?? 'ok',
        value: json['value'] as String?,
        note: json['note'] as String?,
      );
}

class AviationBriefing {
  final String location;
  final String flightCategory;
  final String summary;
  final List<AviationConditionItem> conditions;
  final List<String> bestWindows;

  const AviationBriefing({
    required this.location,
    required this.flightCategory,
    required this.summary,
    required this.conditions,
    required this.bestWindows,
  });

  factory AviationBriefing.fromJson(Map<String, dynamic> json) =>
      AviationBriefing(
        location: json['location'] as String? ?? 'Unknown',
        flightCategory: json['flight_category'] as String? ?? 'unknown',
        summary: json['summary'] as String? ?? '',
        conditions: (json['conditions'] as List<dynamic>?)
                ?.map((e) =>
                    AviationConditionItem.fromJson(e as Map<String, dynamic>))
                .toList() ??
            [],
        bestWindows: (json['best_windows'] as List<dynamic>?)
                ?.map((e) => e as String)
                .toList() ??
            [],
      );
}

// ---------------- City overview ----------------

class CitySnapshot {
  final String name;
  final double? temperature;
  final double? humidity;
  final double? windSpeed;
  final String? condition;
  final int? aqi;

  const CitySnapshot({
    required this.name,
    this.temperature,
    this.humidity,
    this.windSpeed,
    this.condition,
    this.aqi,
  });

  factory CitySnapshot.fromJson(Map<String, dynamic> json) => CitySnapshot(
        name: json['name'] as String? ?? 'Unknown',
        temperature: (json['temperature'] as num?)?.toDouble(),
        humidity: (json['humidity'] as num?)?.toDouble(),
        windSpeed: (json['wind_speed'] as num?)?.toDouble(),
        condition: json['condition'] as String?,
        aqi: (json['aqi'] as num?)?.toInt(),
      );
}

class CityOverview {
  final List<CitySnapshot> cities;

  const CityOverview({required this.cities});

  factory CityOverview.fromJson(Map<String, dynamic> json) => CityOverview(
        cities: (json['cities'] as List<dynamic>?)
                ?.map((e) => CitySnapshot.fromJson(e as Map<String, dynamic>))
                .toList() ??
            [],
      );
}
